README.md
<h1 align="center" >🔑 Auth Module</h1>
<p align="center">Zero-boilerplate authentication support for Nuxt.js!</p>

<p align="center">
<a href="https://auth.nuxtjs.org">Refer to the document before usage(具体使用请参照官方文档)</a>
</p>

## Modify Remark
The origin proect cant't support auth runtimeConfig. 
So i have to build again and again when I deploy the site within different subdomains(eg https://xxx.com/a and https://xxx.com/b), what a waste of time!!!

本项目在原项目基础上实现了cookie与localStorage节点的runtime配置，从而实现“一套代码、多处部署”，即使在同一个域名下也不会互相干扰。
使用示例如下
```
publicRuntimeConfig: {
	auth: {
	  cookie: {
		prefix: 'test.',
		options: {
		  path: '/',
		  maxAge: 2 * 3600
		}
	  },
	  localStorage: {
		prefix: 'test.'
	  }
	}
}
```

## Development

Running demo for development:

```bash
$ yarn install
$ yarn dev
```

Running tests for development:

```bash
$ yarn build
$ yarn nuxt build test/fixture
$ yarn jest
```

## License

[MIT License](./LICENSE) - Copyright (c) Nuxt Community


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/wensaint/auth-module/compare/v1.0.3...v1.0.2) (2022-06-24)

### [1.0.1](https://github.com/wensaint/auth-module/compare/v1.0.3...v1.0.1) (2022-06-24)

### [1.0.3](https://github.com/wensaint/auth-module/compare/v1.0.2...v1.0.3) (2022-06-24)

### [1.0.2](https://github.com/wensaint/auth-module/compare/v1.0.1...v1.0.2) (2022-06-24)

### [1.0.1](https://github.com/wensaint/auth-module/compare/v5.0.0...v1.0.1) (2022-06-24)

## [5.0.0](https://github.com/wensaint/auth-module/compare/v4.9.0...v5.0.0) (2022-06-24)


### ⚠ BREAKING CHANGES

* remove build artifact from repo (#1495)

### Features

* add `Token`, `RefreshToken` and `TokenStatus` classes ([b613533](https://github.com/wensaint/auth-module/commit/b61353363529585ec928bc62ffe86e1d236ba781))
* add `Token`, `RefreshToken` and `TokenStatus` classes ([5e57bd4](https://github.com/wensaint/auth-module/commit/5e57bd4bacf46d4ed92d8183391d39ddfdd3d8e9))
* add discord provider ([#914](https://github.com/wensaint/auth-module/issues/914)) ([5a3c3a8](https://github.com/wensaint/auth-module/commit/5a3c3a8a53195618923726b70f19b2ee8336b333))
* add password grant utility and fix client id and grant type ([#602](https://github.com/wensaint/auth-module/issues/602)) ([f47885c](https://github.com/wensaint/auth-module/commit/f47885c6af83af08506fe05e255107ecf02edb22)), closes [#633](https://github.com/wensaint/auth-module/issues/633)
* auth0 scheme ([8483fb3](https://github.com/wensaint/auth-module/commit/8483fb360030734a7c4f986a5d77459b2804c45c))
* **auth:** support strategy.check ([d53d57b](https://github.com/wensaint/auth-module/commit/d53d57b8998aaa0c271fc70135d4c6ac60918738))
* cookie scheme and laravel sanctum ([6b8fa2b](https://github.com/wensaint/auth-module/commit/6b8fa2bee10276a2b37901fe03e4010002b4e8b7))
* cookie-scheme ([b0bfafb](https://github.com/wensaint/auth-module/commit/b0bfafb5f63c7c286a9725786a872a3eb7a6c1d9))
* **cookie:** keep cookie or reponse code as token ([42f1db8](https://github.com/wensaint/auth-module/commit/42f1db8e2c9419015d380c267d33f1fff445f0d7))
* **core:** add `refreshTokenAt` method ([75374e1](https://github.com/wensaint/auth-module/commit/75374e14edb63f9a882a4c1f447b3520e35966fa))
* **core:** add `refreshTokenAt` method ([e432774](https://github.com/wensaint/auth-module/commit/e43277436b1d3848514ebabb60f903dbfd6628b8))
* **core:** add `scheduleTokenRefresh` method ([5b133c5](https://github.com/wensaint/auth-module/commit/5b133c5e8bc20545c9f848cbe84dee5d47b15caf))
* **core:** add `scheduleTokenRefresh` method ([1c474ad](https://github.com/wensaint/auth-module/commit/1c474adf2f06fa020f7e3fff6101aab7beb63a09))
* **core:** add common token refresh controller ([83cb345](https://github.com/wensaint/auth-module/commit/83cb34567de5148e0e5421ba151f40ade2efc901))
* **core:** add common token refresh controller ([a07c412](https://github.com/wensaint/auth-module/commit/a07c412b03e777e4028eb7ca39979e73dd442f26))
* **core:** add custom exception for expired sessions ([531cc15](https://github.com/wensaint/auth-module/commit/531cc15e04f183eb62a404c7ab794c60cbc5015e))
* **core:** add custom exception for expired sessions ([5f38665](https://github.com/wensaint/auth-module/commit/5f386651ae788d19d394935eb9c75472a8ee0642))
* **core:** add methods `setTokenExpiration` and `setRefreshTokenExpiration` ([5fc7e87](https://github.com/wensaint/auth-module/commit/5fc7e879ccd93b8312c30768601473595a8cfc4a))
* **core:** add methods `setTokenExpiration` and `setRefreshTokenExpiration` ([88bda9b](https://github.com/wensaint/auth-module/commit/88bda9bddf78f2df95f67030878ee31688d3a417))
* **core:** add refresh token method ([6082652](https://github.com/wensaint/auth-module/commit/608265294adbe89dcad9eb8e56e85fa88f4f2305))
* **core:** add token expiration helpers ([a42ba8e](https://github.com/wensaint/auth-module/commit/a42ba8e56347a8ccefc8134c3d435f5fc00ce159))
* **core:** add token expiration helpers ([19aa116](https://github.com/wensaint/auth-module/commit/19aa11610a167c904c0216846bc7a6290e54397f))
* **core:** move refresh code from utilities to its own file ([399f3ab](https://github.com/wensaint/auth-module/commit/399f3ab61719e0579ab2278c70a0129eeca558ed))
* **core:** move refresh code from utilities to its own file ([238864e](https://github.com/wensaint/auth-module/commit/238864e6746579c0521e8e776651f88feaf2c85d))
* **core:** use `Token` and `RefreshToken` classes ([bf84bbf](https://github.com/wensaint/auth-module/commit/bf84bbf608b8a3258a6aa39aae7f0f707e157f1f))
* **core:** use `Token` and `RefreshToken` classes ([30f0194](https://github.com/wensaint/auth-module/commit/30f01949eb22a794dd29e7c6ea1975b7995f7022))
* **defaults:** add `tokenExpiration` and `refreshTokenExpiration` prefixes ([8b23f57](https://github.com/wensaint/auth-module/commit/8b23f57f4fc7f2b139d52735a51bd1b88d62128e))
* **defaults:** add `tokenExpiration` and `refreshTokenExpiration` prefixes ([6142403](https://github.com/wensaint/auth-module/commit/61424033e7543ff547c2399b80c3684f595705a6))
* **demo:** add refresh token button ([2438182](https://github.com/wensaint/auth-module/commit/243818269af740545179b56ef27c7bcfbe851b5f))
* **demo:** improve refresh endpoint ([c4f0440](https://github.com/wensaint/auth-module/commit/c4f04405a55ad5e2ea83ad058123940460589096))
* deprecate `getResponseProp` ([#1006](https://github.com/wensaint/auth-module/issues/1006)) ([3646657](https://github.com/wensaint/auth-module/commit/364665758ef53476c60d945369e470b69981b318)), closes [#952](https://github.com/wensaint/auth-module/issues/952)
* **examples:** add refresh example ([8a3340e](https://github.com/wensaint/auth-module/commit/8a3340ef91cb458af161f3ef5784afb58dcc8da9))

src/core/storage.ts
import type { Context } from '@nuxt/types'
import Vue from 'vue'
import cookie from 'cookie'
import type { ModuleOptions } from '../options'
import { isUnset, isSet, decodeValue, encodeValue, getProp } from '../utils'

// TODO: Normalize type at module itself
export type StorageOptions = ModuleOptions & {
  initialState: {
    user: null
    loggedIn: boolean
  }
}

// TODO: Improve type of storages: Universal / Cookie / Local / State

export class Storage {
  public ctx: Context
  public options: StorageOptions

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public state: any
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private _state: any
  private _useVuex: boolean

  constructor(ctx: Context, options: StorageOptions) {
    this.ctx = ctx
    // this.options = options
    // add runtimeconfig support by wensaint
    this.options = {
      ...options,
      ...(this.ctx.$config && this.ctx.$config.auth)
    }
    // this.options = Object.assign(options, {
    //   ...this.ctx.$config.auth
    // })
    this._initState()
  }

  // ------------------------------------
  // Universal
  // ------------------------------------

  setUniversal<V extends unknown>(key: string, value: V): V | void {
    // Unset null, undefined
    if (isUnset(value)) {
      return this.removeUniversal(key)
    }

    // Cookies
    this.setCookie(key, value)

    // Local Storage
    this.setLocalStorage(key, value)

    // Local state
    this.setState(key, value)

    return value
  }

  getUniversal(key: string): unknown {
    let value

    // Local state
    if (process.server) {
      value = this.getState(key)
    }

    // Cookies
    if (isUnset(value)) {
      value = this.getCookie(key)
    }

    // Local Storage
    if (isUnset(value)) {
      value = this.getLocalStorage(key)
    }

    // Local state
    if (isUnset(value)) {
      value = this.getState(key)
    }

    return value
  }

  syncUniversal(key: string, defaultValue?: unknown): unknown {
    let value = this.getUniversal(key)

    if (isUnset(value) && isSet(defaultValue)) {
      value = defaultValue
    }

    if (isSet(value)) {
      this.setUniversal(key, value)
    }

    return value
  }

  removeUniversal(key: string): void {
    this.removeState(key)
    this.removeLocalStorage(key)
    this.removeCookie(key)
  }

  // ------------------------------------
  // Local state (reactive)
  // ------------------------------------

  _initState(): void {
    // Private state is suitable to keep information not being exposed to Vuex store
    // This helps prevent stealing token from SSR response HTML
    Vue.set(this, '_state', {})

    // Use vuex for local state's if possible
    this._useVuex = this.options.vuex && !!this.ctx.store

    if (this._useVuex) {
      const storeModule = {
        namespaced: true,
        state: () => this.options.initialState,
        mutations: {
          SET(state, payload) {
            Vue.set(state, payload.key, payload.value)
          }
        }
      }

      this.ctx.store.registerModule(this.options.vuex.namespace, storeModule, {
        preserveState: Boolean(
          this.ctx.store.state[this.options.vuex.namespace]
        )
      })

      this.state = this.ctx.store.state[this.options.vuex.namespace]
    } else {
      Vue.set(this, 'state', {})

      // eslint-disable-next-line no-console
      console.warn(
        '[AUTH] The Vuex Store is not activated. This might cause issues in auth module behavior, like redirects not working properly.' +
        'To activate it, see https://nuxtjs.org/docs/2.x/directory-structure/store'
      )
    }
  }

  setState<V extends unknown>(key: string, value: V): V {
    if (key[0] === '_') {
      Vue.set(this._state, key, value)
    } else if (this._useVuex) {
      this.ctx.store.commit(this.options.vuex.namespace + '/SET', {
        key,
        value
      })
    } else {
      Vue.set(this.state, key, value)
    }

    return value
  }

  getState(key: string): unknown {
    if (key[0] !== '_') {
      return this.state[key]
    } else {
      return this._state[key]
    }
  }

  watchState(
    key: string,
    fn: (value: unknown, oldValue: unknown) => void
  ): () => void {
    if (this._useVuex) {
      return this.ctx.store.watch(
        (state) => getProp(state[this.options.vuex.namespace], key),
        fn
      )
    }
  }

  removeState(key: string): void {
    this.setState(key, undefined)
  }

  // ------------------------------------
  // Local storage
  // ------------------------------------

  setLocalStorage<V extends unknown>(key: string, value: V): V | void {
    // Unset null, undefined
    if (isUnset(value)) {
      return this.removeLocalStorage(key)
    }

    if (!this.isLocalStorageEnabled()) {
      return
    }

    const _key = this.getPrefix() + key

    try {
      localStorage.setItem(_key, encodeValue(value))
    } catch (e) {
      if (!this.options.ignoreExceptions) {
        throw e
      }
    }

    return value
  }

  getLocalStorage(key: string): unknown {
    if (!this.isLocalStorageEnabled()) {
      return
    }

    const _key = this.getPrefix() + key

    const value = localStorage.getItem(_key)

    return decodeValue(value)
  }

  removeLocalStorage(key: string): void {
    if (!this.isLocalStorageEnabled()) {
      return
    }

    const _key = this.getPrefix() + key

    localStorage.removeItem(_key)
  }

  // ------------------------------------
  // Cookies
  // ------------------------------------
  getCookies(): Record<string, unknown> {
    if (!this.isCookiesEnabled()) {
      return
    }
    const cookieStr = process.client
      ? document.cookie
      : this.ctx.req.headers.cookie

    return cookie.parse(cookieStr || '') || {}
  }

  setCookie<V extends unknown>(
    key: string,
    value: V,
    options: { prefix?: string } = {}
  ): V {
    if (!this.options.cookie || (process.server && !this.ctx.res)) {
      return
    }

    if (!this.isCookiesEnabled()) {
      return
    }

    const _prefix =
      options.prefix !== undefined ? options.prefix : this.options.cookie.prefix
    const _key = _prefix + key
    const _options = Object.assign({}, this.options.cookie.options, options)
    const _value = encodeValue(value)

    // Unset null, undefined
    if (isUnset(value)) {
      _options.maxAge = -1
    }

    // Accept expires as a number for js-cookie compatiblity
    if (typeof _options.expires === 'number') {
      _options.expires = new Date(Date.now() + _options.expires * 864e5)
    }

    const serializedCookie = cookie.serialize(_key, _value, _options)

    if (process.client) {
      // Set in browser
      document.cookie = serializedCookie
    } else if (process.server && this.ctx.res) {
      // Send Set-Cookie header from server side
      let cookies = (this.ctx.res.getHeader('Set-Cookie') as string[]) || []
      if (!Array.isArray(cookies)) cookies = [cookies]
      cookies.unshift(serializedCookie)
      this.ctx.res.setHeader(
        'Set-Cookie',
        cookies.filter(
          (v, i, arr) =>
            arr.findIndex((val) =>
              val.startsWith(v.substr(0, v.indexOf('=')))
            ) === i
        )
      )
    }

    return value
  }

  getCookie(key: string): unknown {
    if (!this.options.cookie || (process.server && !this.ctx.req)) {
      return
    }

    if (!this.isCookiesEnabled()) {
      return
    }

    const _key = this.options.cookie.prefix + key

    const cookies = this.getCookies()

    const value = cookies[_key]
      ? decodeURIComponent(cookies[_key] as string)
      : undefined

    return decodeValue(value)
  }

  removeCookie(key: string, options?: { prefix?: string }): void {
    this.setCookie(key, undefined, options)
  }

  getPrefix(): string {
    if (!this.options.localStorage) {
      throw new Error('Cannot get prefix; localStorage is off')
    }
    return this.options.localStorage.prefix
  }

  isLocalStorageEnabled(): boolean {
    // Disabled by configuration
    if (!this.options.localStorage) {
      return false
    }

    // Local Storage only exists in the browser
    if (!process.client) {
      return false
    }

    // There's no great way to check if localStorage is enabled; most solutions
    // error out. So have to use this hacky approach :\
    // https://stackoverflow.com/questions/16427636/check-if-localstorage-is-available
    const test = 'test'
    try {
      localStorage.setItem(test, test)
      localStorage.removeItem(test)
      return true
    } catch (e) {
      if (!this.options.ignoreExceptions) {
        // eslint-disable-next-line no-console
        console.warn(
          "[AUTH] Local storage is enabled in config, but browser doesn't" +
          ' support it'
        )
      }
      return false
    }
  }

  isCookiesEnabled(): boolean {
    // Disabled by configuration
    if (!this.options.cookie) {
      return false
    }

    // Server can only assume cookies are enabled, it's up to the client browser
    // to create them or not
    if (process.server) {
      return true
    }

    if (window.navigator.cookieEnabled) {
      return true
    } else {
      // eslint-disable-next-line no-console
      console.warn(
        "[AUTH] Cookies is enabled in config, but browser doesn't" +
        ' support it'
      )
      return false
    }
  }
}


demo/nuxt.config.ts
import { NuxtConfig } from '@nuxt/types'
import oidcMockServer from './api/oidcmockserver'

export default <NuxtConfig>{
  build: {
    extractCSS: true
  },
  serverMiddleware: [
    '~/api/auth',
    '~/api/oauth2mockserver',
    {
      path: '/oidc',
      handler: oidcMockServer({
        port: 3000,
        path: '/oidc',
        redirect: {
          callback: '/callback'
        }
      })
    }
  ],
  buildModules: ['@nuxt/typescript-build'],
  modules: ['bootstrap-vue/nuxt', '@nuxtjs/axios', '../src/module'],
  components: true,
  axios: {
    proxy: true
  },
  proxy: {
    '/api': 'http://localhost:3000',
    '/laravel': {
      target: 'https://laravel-auth.nuxtjs.app',
      pathRewrite: { '^/laravel': '/' }
    }
  },
  auth: {
    redirect: {
      callback: '/callback',
      logout: '/signed-out'
    },
    strategies: {
      local: {
        token: {
          property: 'token.accessToken'
        }
      },
      localRefresh: {
        scheme: 'refresh',
        token: {
          property: 'token.accessToken',
          maxAge: 15
        },
        refreshToken: {
          property: 'token.refreshToken',
          data: 'refreshToken',
          maxAge: false
        }
      },
      auth0: {
        domain: 'nuxt-auth.auth0.com',
        clientId: 'q8lDHfBLJ-Fsziu7bf351OcYQAIe3UJv'
      },
      discord: {
        clientId: process.env.DISCORD_CLIENT_ID,
        clientSecret: process.env.DISCORD_CLIENT_SECRET
      },
      facebook: {
        endpoints: {
          userInfo:
            'https://graph.facebook.com/v2.12/me?fields=about,name,picture{url},email,birthday'
        },
        clientId: '1671464192946675',
        scope: ['public_profile', 'email', 'user_birthday']
      },
      google: {
        clientId:
          '956748748298-kr2t08kdbjq3ke18m3vkl6k843mra1cg.apps.googleusercontent.com'
      },
      github: {
        clientId: process.env.GITHUB_CLIENT_ID,
        clientSecret: process.env.GITHUB_CLIENT_SECRET
      },
      // twitter: {
      //   clientId: 'FAJNuxjMTicff6ciDKLiZ4t0D'
      // },
      laravelJWT: {
        url: '/laravel',
        endpoints: {
          login: {
            url: '/api/auth/jwt/login'
          },
          refresh: {
            url: '/api/auth/jwt/refresh'
          },
          logout: {
            url: '/api/auth/jwt/logout'
          },
          user: {
            url: '/api/auth/jwt/user'
          }
        }
      },
      laravelSanctum: {
        url: '/laravel'
      },
      laravelPassport: {
        url: 'https://laravel-auth.nuxtjs.app',
        endpoints: {
          userInfo: '/api/auth/passport/user'
        },
        token: {
          maxAge: 1800
        },
        refreshToken: {
          maxAge: 60 * 60 * 24 * 30
        },
        clientId: '3',
        clientSecret: 'k0NAhYGKXbG6NjENFz4VIe5YSbccZWW9V3gGeSOa'
      },
      laravelPassportPasswordGrant: {
        name: 'laravelPassportPassword',
        provider: 'laravelPassport',
        url: '/laravel',
        endpoints: {
          user: {
            url: '/api/auth/passport/user'
          }
        },
        token: {
          maxAge: 1800
        },
        refreshToken: {
          maxAge: 60 * 60 * 24 * 30
        },
        clientId: '2',
        clientSecret: 'eKm1ei8muaql7TfcBxhN6Nq48oSflw6QJKCZF8gl',
        grantType: 'password'
      },
      oauth2mock: {
        scheme: 'oauth2',
        endpoints: {
          authorization: '/oauth2mockLogin',
          token: '/oauth2mockserver/token',
          userInfo: '/oauth2mockserver/userinfo'
        },
        responseType: 'code',
        grantType: 'authorization_code',
        clientId: 'test-client'
      },
      oidcmock: {
        scheme: 'openIDConnect',
        responseType: 'code',
        scope: ['openid', 'profile', 'offline_access'],
        grantType: 'authorization_code',
        clientId: 'oidc_authorization_code_client',
        logoutRedirectUri: 'http://localhost:3000',
        endpoints: {
          configuration:
            'http://localhost:3000/oidc/.well-known/openid-configuration'
        }
      }
    }
  },
  publicRuntimeConfig: {
    auth: {
      cookie: {
        prefix: 'test.',
        options: {
          path: '/',
          maxAge: 2 * 3600
        }
      },
      localStorage: {
        prefix: 'test.'
      }
    }
  }
}
